warehouse/schema/schema.go

```diff
@@ -130,12 +133,138 @@ func (sh *Schema) fetchSchemaFromWarehouse(ctx context.Context, repo fetchSchema
 	sh.removeDeprecatedColumns(warehouseSchema)
 	sh.removeDeprecatedColumns(unrecognizedWarehouseSchema)
 
+	sh.schemaInWarehouseMu.Lock()
+	defer sh.schemaInWarehouseMu.Unlock()
+	sh.unrecognizedSchemaInWarehouseMu.Lock()
+	defer sh.unrecognizedSchemaInWarehouseMu.Unlock()
+
 	sh.schemaInWarehouse = warehouseSchema
 	sh.unrecognizedSchemaInWarehouse = unrecognizedWarehouseSchema
 
 	return nil
 }
 
+func (sh *Schema) SyncRemoteSchema(ctx context.Context, fetchSchemaRepo fetchSchemaRepo, uploadID int64) (bool, error) {
+	localSchema, err := sh.GetLocalSchema(ctx)
+	if err != nil {
+		return false, fmt.Errorf("fetching schema from local: %w", err)
+	}
+	if err := sh.FetchSchemaFromWarehouse(ctx, fetchSchemaRepo); err != nil {
+		return false, fmt.Errorf("fetching schema from warehouse: %w", err)
+	}
+
+	sh.schemaInWarehouseMu.RLock()
+	schemaChanged := sh.hasSchemaChanged(localSchema)
+	defer sh.schemaInWarehouseMu.RUnlock()
```

Comment:
[Optional] Maybe we can simplify this. Currently, we are using `RUnlock` in a defer block only if schemaChanged.
```suggestion
	sh.schemaInWarehouseMu.RLock()
	defer sh.schemaInWarehouseMu.RUnlock()
	schemaChanged := sh.hasSchemaChanged(localSchema)
```

Reply:
Good catch! :+1: I think I know the keyboard shortcut responsible for this :sweat_smile: Thanks :+1: